src/distrs/binom.jl

```diff
@@ -16,4 +16,13 @@ import .RFunctions:
 binompdf(n::Real, p::Real, k::Real) = exp(binomlogpdf(n, p, k))
 
 # logpdf for numbers with generic types
-binomlogpdf(n::Real, p::Real, k::Real) = -log1p(n) - logbeta(n - k + 1, k + 1) + k * log(p) + (n - k) * log1p(-p)
+function binomlogpdf(n::Real, p::Real, k::Real)
+    if isinteger(k) & (zero(k) <= k <= n)
+        x = loggamma(n + 1) - loggamma(k + 1) - loggamma(n - k + 1) +
+            xlogy(k, p) + xlogy(n - k, 1 - p)
+    else
+        x = -Inf
+    end
+
+    return convert(float(promote_type(typeof(n), typeof(p), typeof(k))), x)
```

Comment:
Generally, it's better to just compute as much of the general base as possible, i.e. without risking exceptions, and then use the type of that result as the return type. It would be something like
```julia
_lpdf = xlogy(k, p) + xlogy(n - k, 1 - p)
if isinteger(k) & (zero(k) <= k <= n)
    return _lpdf + loggamma(n + 1) - loggamma(k + 1) - loggamma(n - k + 1)
else
    return oftype(_lpdf, -Inf)
end
```
Another thing, by computing `1-p` explicitly, we are losing some precision when `n` is very large. Maybe we need a `xlog1py` or we should just special case the corner case here in this function and continue to use `log1p`.

Reply:
Is that really better? I mean, it's much more complex than `float` and it's still not completely right. If we go to that level of complexity, let's do it 100% correctly by also calling `loggamma(one(n-k))`?

Reply:
Yes. It's better to avoid `promote_type` and just use a computed value of the correct type. I can't come up with a case where `xlogy` and `loggamma` would produce different types.

Reply:
Fine with me -- but I find it really annoying that we have to rely on soft rules like that... :-/

Reply:
Yes. I've also been annoyed by this. I've been wondering if you could have some kind of language support for branches that enforces the same return type.

Reply:
Yes, something like `@sametype if` would be nice. But that would only work if inference is able to figure out the type, and AFAIK we shouldn't rely on it. What we could rely on is the fact that for `Number` variables, `one` is defined: the macro could compute the first branch with all variables set to `one(var)`, and then convert the result of the second branch to the type of the result. We probably need a more specific name for that macro...